File: frontend/src/types/agent.ts
// ── Enums ─────────────────────────────────────────────────────────────────────

export type AgentMode = 'single' | 'crew'
export type MemoryType = 'none' | 'session' | 'persistent' | 'summary'
export type CrewProcess = 'sequential' | 'hierarchical' | 'parallel'
export type AutonomyLevel = 'reactive' | 'semi' | 'autonomous'
export type ChannelType = 'web_chat' | 'whatsapp' | 'telegram' | 'slack' | 'rest_api'
export type BuildStatus = 'pending' | 'building' | 'ready' | 'failed'
export type AgentStatus = 'draft' | 'building' | 'testing' | 'deployed' | 'archived'


// ── Sub-specs ─────────────────────────────────────────────────────────────────

export interface ToolRef {
  name: string
  source: 'library' | 'custom'
  config: Record<string, unknown>
}

export interface MemorySpec {
  type: MemoryType
  ttl_seconds?: number
  max_messages: number
}

export interface RAGSpec {
  enabled: boolean
  pack: string
  sources: string[]
  chunk_size: number
  chunk_overlap: number
  top_k: number
  embedding_model: string
}

export interface ModelParams {
  model: string
  provider?: string
  base_url?: string
  temperature: number
  max_tokens: number
  top_p: number
  llm_key_id?: string
}

export interface AgentRoleSpec {
  name: string
  role: string
  goal: string
  backstory: string
  tools: ToolRef[]
  allow_delegation: boolean
  model_params?: ModelParams
}


// ── AgentSpec ─────────────────────────────────────────────────────────────────

export interface AgentSpec {
  tenant_id: string
  name: string
  description: string
  goal: string
  mode: AgentMode
  channels: ChannelType[]
  model_params: ModelParams
  constraints: string[]
  expected_inputs: string[]
  expected_outputs: string[]
  memory: MemorySpec
  rag: RAGSpec
  // single
  tools: ToolRef[]
  autonomy_level: AutonomyLevel
  // crew
  agents: AgentRoleSpec[]
  process: CrewProcess
  manager_model?: string
}


// ── Design output ─────────────────────────────────────────────────────────────

export interface FrameworkSelection {
  framework: 'langchain' | 'crewai'
  agent_type?: 'openai_functions' | 'react'
  process?: CrewProcess
  justification: string
  estimated_complexity: 'low' | 'medium' | 'high'
}

export type FlowNodeType = 'start' | 'agent' | 'decision' | 'tool' | 'end'

export interface FlowNodePosition {
  x: number
  y: number
}

export interface FlowNodeData {
  description?: string
  assigned_agent?: string
  allowed_tools?: string[]
  tool_name?: string
  notes?: string
}

export interface FlowNode {
  id: string
  type: FlowNodeType
  label: string
  description: string
  position?: FlowNodePosition
  data?: FlowNodeData
}

export interface FlowEdge {
  id: string
  from: string
  to: string
  condition?: string | null
}

export interface GraphMeta {
  version: number
  layout: 'manual' | 'auto'
}

export interface GraphBlueprint {
  nodes: FlowNode[]
  edges: FlowEdge[]
  meta?: GraphMeta
}

export interface GraphValidationIssue {
  level: 'error' | 'warning'
  code: string
  message: string
  node_id?: string | null
  edge_id?: string | null
}

export interface GraphValidationReport {
  ok: boolean
  errors: GraphValidationIssue[]
  warnings: GraphValidationIssue[]
}

export interface GraphUpdateResponse {
  graph_blueprint: GraphBlueprint
  mermaid_diagram: string
  validation: GraphValidationReport
}

export interface AgentDesign {
  agent_id: string
  tenant_id: string
  spec: AgentSpec
  framework: FrameworkSelection
  system_prompt: string
  graph_blueprint: GraphBlueprint
  test_cases: Array<{
    id: number
    description: string
    input: string
    expected_behavior: string
    expected_tools: string[]
    pass_criteria: string
  }>
  mermaid_diagram: string
  version: number
  status?: string  // draft | building | testing | deployed | archived
}

export interface AgentResponse {
  output: string
  session_id?: string
  metadata?: Record<string, unknown>
}


// ── Wizard state (frontend) ───────────────────────────────────────────────────

export interface WizardState {
  step: number
  mode?: AgentMode
  name: string
  description: string
  goal: string
  channels: ChannelType[]
  tools: ToolRef[]
  memory: MemorySpec
  rag: RAGSpec
  model_params: ModelParams
  constraints: string[]
  autonomy_level: AutonomyLevel
  // crew
  agents: AgentRoleSpec[]
  process: CrewProcess
}

export interface WizardAdvisorItem {
  title: string
  detail: string
  severity: 'info' | 'warning' | 'critical'
}

export interface WizardAdvisorRequest {
  step: number
  step_key: string
  final_review?: boolean
  wizard_state: WizardState
  available_tools?: AvailableToolDescriptor[]
  tool_readiness?: ToolReadinessStatus[]
  available_models?: Array<Record<string, unknown>>
}

export interface WizardAdvisorResponse {
  score: number
  status: 'ready' | 'requires_review' | 'high_risk'
  summary: string
  suggestions: WizardAdvisorItem[]
  risks: WizardAdvisorItem[]
  questions: WizardAdvisorItem[]
  proposed_patch: Partial<WizardState>
  source: 'ai' | 'rules'
}

export const WIZARD_DEFAULTS: WizardState = {
  step: 0,
  name: '',
  description: '',
  goal: '',
  channels: ['web_chat'],
  tools: [],
  memory: { type: 'session', max_messages: 50 },
  rag: {
    enabled: false,
    pack: 'GenericDocsRAG',
    sources: [],
    chunk_size: 500,
    chunk_overlap: 50,
    top_k: 4,
    embedding_model: 'text-embedding-3-small',
  },
  model_params: {
    model: 'claude-sonnet-4-5',
    temperature: 0.3,
    max_tokens: 2048,
    top_p: 1.0,
  },
  constraints: [],
  autonomy_level: 'reactive',
  agents: [],
  process: 'sequential',
}


// ── Skills ────────────────────────────────────────────────────────────────────

export interface Skill {
  id: string
  name: string
  description: string
  objective: string
  usage_conditions: string
  tools: ToolRef[]
  procedure: string
  quality_rules: string
  output_format: string
  guardrails: string[]
  is_active: boolean
  created_at: string
}

// ── MCP Servers ───────────────────────────────────────────────────────────────

export interface DiscoveredTool {
  name: string
  description: string
  input_schema: Record<string, unknown>
}

export interface MCPServer {
  id: string
  name: string
  endpoint: string
  transport: 'sse' | 'http'
  auth_type: 'none' | 'bearer' | 'basic'
  auth_config: Record<string, string>
  discovered_tools: DiscoveredTool[]
  is_active: boolean
  last_tested_at?: string
}

// ── Knowledge Bases ───────────────────────────────────────────────────────────

export interface KnowledgeBase {
  id: string
  name: string
  description: string
  rag_spec: RAGSpec
  status: 'empty' | 'indexing' | 'ready' | 'error'
  created_at: string
}

// ── Behavior Policy ───────────────────────────────────────────────────────────

export interface BehaviorPolicy {
  id?: string
  agent_id?: string
  tone: string
  escalation_conditions: string[]
  confirmation_triggers: string[]
  format_requirements: string
  custom_rules: string[]
}

// ── Guardrail Rules ───────────────────────────────────────────────────────────

export type GuardrailRuleType = 'input_block' | 'output_filter' | 'length_limit' | 'topic_restrict'
export type GuardrailAction = 'block' | 'warn' | 'transform'

export interface GuardrailRule {
  id?: string
  agent_id?: string
  name: string
  rule_type: GuardrailRuleType
  condition: Record<string, unknown>
  action: GuardrailAction
  priority: number
  is_active: boolean
}

// ── Available tools en la Component Library ──────────────────────────────────

export type ToolOperationalState = 'ready' | 'needs_config' | 'framework_limited' | 'disabled'
export type ToolFrameworkState = 'ready' | 'limited' | 'unsupported'
export type ToolFramework = 'langchain' | 'crewai'

export interface ToolReadinessStatus {
  name: string
  configured: boolean
  state: ToolOperationalState
  state_label: string
  setup_hint?: string
}

export interface AvailableToolDescriptor {
  name: string
  description: string
  category: string
  state: ToolOperationalState
  state_label: string
  setup_hint?: string
  frameworks: {
    langchain: ToolFrameworkState
    crewai: ToolFrameworkState
  }
}

export function getToolFrameworkState(toolName: string, framework: ToolFramework): ToolFrameworkState {
  return AVAILABLE_TOOLS.find(tool => tool.name === toolName)?.frameworks[framework] ?? 'unsupported'
}

export function isToolSupportedInFramework(toolName: string, framework: ToolFramework): boolean {
  return getToolFrameworkState(toolName, framework) !== 'unsupported'
}

export function applyToolReadiness(
  tool: AvailableToolDescriptor,
  readinessMap?: Record<string, ToolReadinessStatus>
): AvailableToolDescriptor {
  const readiness = readinessMap?.[tool.name]
  if (!readiness) return tool
  return {
    ...tool,
    state: readiness.state,
    state_label: readiness.state_label,
    setup_hint: readiness.setup_hint,
  }
}

export const AVAILABLE_TOOLS: AvailableToolDescriptor[] = [
  {
    name: 'web_search',
    description: 'Buscar información en internet (Tavily)',
    category: 'información',
    state: 'needs_config',
    state_label: 'Requiere credencial',
    setup_hint: 'Necesita una clave Tavily configurada en el backend o resuelta desde configuración global.',
    frameworks: { langchain: 'ready', crewai: 'ready' },
  },
  {
    name: 'sql_query',
    description: 'Consultar bases de datos SQL',
    category: 'datos',
    state: 'needs_config',
    state_label: 'Requiere datasource',
    setup_hint: 'Necesita un DSN o conexión segura y todavía no tiene una UX completa de configuración.',
    frameworks: { langchain: 'ready', crewai: 'ready' },
  },
  {
    name: 'rest_api_call',
    description: 'Llamar APIs REST externas',
    category: 'integración',
    state: 'needs_config',
    state_label: 'Requiere política',
    setup_hint: 'Conviene definir dominios permitidos, métodos y headers por defecto antes de usarla en producción.',
    frameworks: { langchain: 'ready', crewai: 'ready' },
  },
  {
    name: 'calculator',
    description: 'Evaluar expresiones matemáticas',
    category: 'utilidad',
    state: 'ready',
    state_label: 'Lista',
    setup_hint: 'No requiere credenciales ni configuración adicional.',
    frameworks: { langchain: 'ready', crewai: 'ready' },
  },
  {
    name: 'send_email',
    description: 'Enviar emails via SMTP',
    category: 'comunicación',
    state: 'needs_config',
    state_label: 'Requiere SMTP global',
    setup_hint: 'Depende de una configuración SMTP global válida en la plataforma; no requiere credenciales separadas por tool.',
    frameworks: { langchain: 'ready', crewai: 'ready' },
  },
]

export const AVAILABLE_MODELS: Array<{ id: string; name: string; provider: string; providerId: string }> = [
  { id: 'claude-sonnet-4-5',        name: 'Claude Sonnet 4.5',  provider: 'Anthropic', providerId: 'anthropic' },
  { id: 'claude-opus-4-6',          name: 'Claude Opus 4.6',    provider: 'Anthropic', providerId: 'anthropic' },
  { id: 'claude-haiku-4-5-20251001',name: 'Claude Haiku 4.5',   provider: 'Anthropic', providerId: 'anthropic' },
  { id: 'gpt-4o',                   name: 'GPT-4o',             provider: 'OpenAI',    providerId: 'openai' },
  { id: 'gpt-4o-mini',              name: 'GPT-4o Mini',        provider: 'OpenAI',    providerId: 'openai' },
  { id: 'openrouter:openai/gpt-4o-mini',           name: 'GPT-4o Mini',       provider: 'OpenRouter', providerId: 'openrouter' },
  { id: 'openrouter:openai/gpt-4o',               name: 'GPT-4o',            provider: 'OpenRouter', providerId: 'openrouter' },
  { id: 'openrouter:anthropic/claude-3.5-sonnet', name: 'Claude 3.5 Sonnet', provider: 'OpenRouter', providerId: 'openrouter' },
  { id: 'openrouter:anthropic/claude-3-haiku',    name: 'Claude 3 Haiku',    provider: 'OpenRouter', providerId: 'openrouter' },
  { id: 'openrouter:meta-llama/llama-3.1-8b-instruct:free', name: 'Llama 3.1 8B (free)', provider: 'OpenRouter', providerId: 'openrouter' },
  { id: 'openrouter:deepseek/deepseek-chat',      name: 'DeepSeek Chat',     provider: 'OpenRouter', providerId: 'openrouter' },
  { id: 'deepseek-chat',                         name: 'DeepSeek Chat',      provider: 'DeepSeek',   providerId: 'deepseek' },
  { id: 'qwen-plus',                             name: 'Qwen Plus',          provider: 'Qwen',       providerId: 'qwen' },
  { id: 'moonshot-v1-8k',                        name: 'Kimi Moonshot',      provider: 'Moonshot',   providerId: 'moonshot' },
  { id: 'glm-4-plus',                            name: 'GLM-4 Plus',         provider: 'Zhipu',      providerId: 'zhipu' },
]

File: backend/app/schemas/agent.py
from __future__ import annotations
from enum import Enum
from typing import Literal, Optional
from uuid import UUID, uuid4

from pydantic import BaseModel, Field


# ── Enumeraciones ────────────────────────────────────────────────────────────

class AgentMode(str, Enum):
    single = "single"   # Un agente con tools → LangChain
    crew   = "crew"     # Equipo de agentes con roles → CrewAI


class MemoryType(str, Enum):
    none          = "none"
    session       = "session"       # Redis, se borra al cerrar sesión
    persistent    = "persistent"    # PostgreSQL, persiste entre sesiones
    summary       = "summary"       # Summary buffer sobre persistent


class CrewProcess(str, Enum):
    sequential   = "sequential"
    hierarchical = "hierarchical"
    parallel     = "parallel"


class AutonomyLevel(str, Enum):
    reactive   = "reactive"    # Solo responde cuando lo invocan
    semi       = "semi"        # Puede hacer follow-up, pide confirmación en acciones
    autonomous = "autonomous"  # Ejecuta sin pedir confirmación


class ChannelType(str, Enum):
    web_chat  = "web_chat"
    whatsapp  = "whatsapp"
    telegram  = "telegram"
    slack     = "slack"
    rest_api  = "rest_api"


# ── Sub-specs ────────────────────────────────────────────────────────────────

class ToolRef(BaseModel):
    """Referencia a una tool de la Component Library o custom."""
    name: str
    source: Literal["library", "custom"] = "library"
    config: dict = Field(default_factory=dict)  # parámetros específicos de la tool


class MemorySpec(BaseModel):
    type: MemoryType = MemoryType.session
    ttl_seconds: Optional[int] = None           # None = sin expiración
    max_messages: int = 50                       # ventana de contexto en memoria


class RAGSpec(BaseModel):
    enabled: bool = False
    pack: str = "GenericDocsRAG"                # nombre del RAG pack
    sources: list[str] = Field(default_factory=list)  # URLs, paths, DSNs
    chunk_size: int = 500
    chunk_overlap: int = 50
    top_k: int = 4                               # fragmentos a recuperar por query
    embedding_model: str = "text-embedding-3-small"


class ModelParams(BaseModel):
    model: str = "claude-3-5-sonnet-20241022"  # o gpt-4o etc
    provider: Optional[str] = None
    base_url: Optional[str] = None
    temperature: float = Field(0.3, ge=0.0, le=1.0)
    max_tokens: int = Field(2048, ge=256, le=8192)
    top_p: float = Field(1.0, ge=0.0, le=1.0)
    llm_key_id: Optional[str] = None


class FlowNodeType(str, Enum):
    start = "start"
    agent = "agent"
    decision = "decision"
    tool = "tool"
    end = "end"


class FlowNodePosition(BaseModel):
    x: float = 0.0
    y: float = 0.0


class FlowNodeData(BaseModel):
    description: Optional[str] = None
    assigned_agent: Optional[str] = None
    allowed_tools: list[str] = Field(default_factory=list)
    tool_name: Optional[str] = None
    notes: Optional[str] = None


class FlowNode(BaseModel):
    id: str
    type: FlowNodeType
    label: str
    description: str = ""
    position: Optional[FlowNodePosition] = None
    data: FlowNodeData = Field(default_factory=FlowNodeData)


class FlowEdge(BaseModel):
    id: Optional[str] = None
    from_: str = Field(alias="from")
    to: str
    condition: Optional[str] = None

    model_config = {"populate_by_name": True}

    def model_dump(self, *args, **kwargs):
        kwargs.setdefault("by_alias", True)
        return super().model_dump(*args, **kwargs)


class GraphMeta(BaseModel):
    version: int = 1
    layout: Literal["manual", "auto"] = "manual"


class GraphBlueprint(BaseModel):
    nodes: list[FlowNode] = Field(default_factory=list)
    edges: list[FlowEdge] = Field(default_factory=list)
    meta: GraphMeta = Field(default_factory=GraphMeta)

    def as_dict(self) -> dict:
        return self.model_dump(by_alias=True)


class GraphValidationIssue(BaseModel):
    level: Literal["error", "warning"]
    code: str
    message: str
    node_id: Optional[str] = None
    edge_id: Optional[str] = None


class GraphValidationReport(BaseModel):
    ok: bool
    errors: list[GraphValidationIssue] = Field(default_factory=list)
    warnings: list[GraphValidationIssue] = Field(default_factory=list)


class GraphUpdateRequest(BaseModel):
    graph_blueprint: GraphBlueprint
    auto_regenerate_mermaid: bool = True


class GraphUpdateResponse(BaseModel):
    graph_blueprint: GraphBlueprint
    mermaid_diagram: str
    validation: GraphValidationReport


# ── AgentRoleSpec (solo para mode=crew) ─────────────────────────────────────

class AgentRoleSpec(BaseModel):
    model_config = {"protected_namespaces": ()}

    name: str                         # identificador interno, ej: "researcher"
    role: str                         # título visible, ej: "Investigador Senior"
    goal: str                         # objetivo del rol
    backstory: str                    # contexto del rol para CrewAI
    tools: list[ToolRef] = Field(default_factory=list)
    allow_delegation: bool = False
    model_params: Optional[ModelParams] = None  # None = usa el del Crew


# ── AgentSpec — el objeto central del Requirement Wizard ────────────────────

class AgentSpec(BaseModel):
    model_config = {"protected_namespaces": ()}

    """
    Producido por el Requirement Wizard.
    Es el único input que necesita el resto del pipeline.
    """
    # Identidad
    tenant_id: str
    name: str
    description: str
    goal: str

    # Modo — determina el framework
    mode: AgentMode

    # Canales de despliegue
    channels: list[ChannelType] = Field(default_factory=lambda: [ChannelType.web_chat])

    # Configuración del modelo base
    model_params: ModelParams = Field(default_factory=ModelParams)

    # Restricciones y contexto
    constraints: list[str] = Field(default_factory=list)
    expected_inputs: list[str] = Field(default_factory=list)
    expected_outputs: list[str] = Field(default_factory=list)

    # Memoria y RAG
    memory: MemorySpec = Field(default_factory=MemorySpec)
    rag: RAGSpec = Field(default_factory=RAGSpec)

    # Solo para mode=single
    tools: list[ToolRef] = Field(default_factory=list)
    autonomy_level: AutonomyLevel = AutonomyLevel.reactive

    # Solo para mode=crew
    agents: list[AgentRoleSpec] = Field(default_factory=list)
    process: CrewProcess = CrewProcess.sequential
    manager_model: Optional[str] = None   # para process=hierarchical

    # Referencias a entidades de la librería (asignadas post-creación)
    skill_ids: list[str] = Field(default_factory=list)
    mcp_server_ids: list[str] = Field(default_factory=list)
    knowledge_base_ids: list[str] = Field(default_factory=list)


# ── FrameworkSelection — output del Framework Selector ──────────────────────

class FrameworkSelection(BaseModel):
    framework: Literal["langchain", "crewai"]
    agent_type: Optional[Literal["openai_functions", "react"]] = None  # solo langchain
    process: Optional[CrewProcess] = None                               # solo crewai
    justification: str
    estimated_complexity: Literal["low", "medium", "high"]


# ── AgentDesign — output del Design Generator ───────────────────────────────

class AgentDesign(BaseModel):
    schema_version: int = 1
    agent_id: UUID = Field(default_factory=uuid4)
    tenant_id: str
    spec: AgentSpec
    framework: FrameworkSelection
    system_prompt: str
    graph_blueprint: dict = Field(default_factory=dict)  # nodos y edges del grafo
    test_cases: list[dict] = Field(default_factory=list) # mín 5, generados por LLM
    mermaid_diagram: str = ""
    version: int = 1
    status: str = "draft"  # draft | building | testing | deployed | archived


# ── AgentResponse — output unificado del AgentRuntime ───────────────────────

class AgentResponse(BaseModel):
    output: str
    steps: list[dict] = Field(default_factory=list)  # intermediate steps
    tokens_in: int = 0
    tokens_out: int = 0
    latency_ms: float = 0.0
    session_id: str = ""
